### ISSUE TITLE
Support album art using MPD's (new) protocol extension

### ISSUE THREAD

**Issue by hdak1945git** [CDrummond/cantata#1400]
**Is your feature request related to a problem? Please describe.**
Tangentially related: I have an album name with a pound/number sign (`#`) embedded in the middle of it. Unfortunately, cantata doesn't seem to properly escape these path names, or simply truncates it, when it requests album art from my web server. I later confirmed this by watching my server logs for the requested album art while cantata was running.

**Describe the solution you'd like**
Beyond fixing the above, a nicer solution would be cut out the middle man altogether, by implementing the new `albumart` command recently added to MPD and [described here](https://github.com/MusicPlayerDaemon/MPD/blob/master/doc/protocol.rst#id172). This would negate the need to run a separate HTTP server (for the sole use of album art at least).

**Describe alternatives you've considered**
Renaming the files would be a workaround -- but it wouldn't fix the URL encoding bug.

**Additional context**
I know the Android client M.A.L.P. supports this feature (and I have tested it--it works), their project page is located over at [Gitlab](https://gitlab.com/gateship-one/malp) (GPLv3) which may serve as a useful reference.
Arch Linux recently pushed a stable version of the mpd (0.21.1-1 in their repository) which supports the `albumart` command.

**Comment by CDrummond** [comment]
Don't have MPD 0.21, and meson build failed for me. However, I have attempted to implement the protocol - probably bugs seeing as I can't test.

Can you checkout git and compile, and test for me? If so, the change is in the `1400-mpd-covers` branch.

If covers still fail, which to be honest the probably will do, can you run Cantata with debug? e.g.

`./cantata --debug=mpd,covers`

**Comment by hdak1945git** [comment]
Managed to get it running, it did not work as you implied; but I was able to fix some of it:

* `supportsCoverDownload()` was missing the `return` statement.
* `MPDConnection::getCover()` had some faulty logic validating chunk size.

Here is a patch containing my attempt: https://gist.github.com/hdak1945git/32197030720ffa5fc9fc7cfebc303bc3

However there are still some bugs I can't track down. The album art downloads fine according to the debug statements, but not all the images appear in the playlist (save for a maybe 2 or 3 albums).  I don't know enough about Qt to debug it,  but I think `emit albumArt` code is failing, since my debug statements aren't appearing. Though, if I double-click on a song with missing album art, cantata will retry fetching it and succeed every time.

Hope that helps somewhat--

**Comment by CDrummond** [comment]
I've update the branch, with changes taken from your diff. Please dont update DBUG statements in patch - makes it hard to see what the actual changes were. But, **thanks** for taking the time to look into this.

Can you update and try again?

**Comment by hdak1945git** [comment]
Updated as requested--still exhibiting the problems mentioned above...

Luckily it seems like a simple change was in order, the "job queue" (for lack of a better phrase) seemed to be be setup incorrectly. Simply put, a line in `CoverDownloader::downloadViaMpd(...)` invokes `jobs.insert(nullptr, job)`, which overwrites any existing job entry. `CoverDownloader::mpdAlbumArt(...)` later on attempts to verify the job it's working on using `findJob(...)` which will fail most of the time because the fields it compares are likely completely different due to the aforementioned insert.

I've made another patch for you to review: https://gist.github.com/hdak1945git/5c1335f9599296965a8a75fdc7708d54 it's based on your latest push (e039683be75b58bd6163926a485639ab11b0cd00) I've tried to mirror what was intended and refrain from modifying existing `DBUG` messages.

Feature wise at least, the above patch and recent commits fulfills what I need, all my album art is served locally using MPD 🎉 (Save a few I found were missing in testing...)
